src/form/input/Input.tsx

```diff
@@ -109,21 +141,24 @@ function Input(props: InputProps) {
   function handleChange(event: React.SyntheticEvent<HTMLInputElement>) {
     if (isNumberInput) {
       const {value: newValue} = event.currentTarget;
+      let formattedNewValue = parseNumber(newValue, locale);
 
-      let formattedNewValue = newValue
-        .replace(NOT_INTEGER_FIRST_CHARACTER_OF_STRING_REGEX, "")
-        .replace(NOT_NUMBER_NOR_DECIMAL_POINT_REGEX, "")
-        .replace(PRECISION_REGEX, "$1");
-
-      if (maxFractionDigits > 0) {
-        const decimalNumberParts = newValue.split(DECIMAL_NUMBER_SEPARATOR);
+      if (
+        IS_LAST_CHARACTER_DECIMAL_POINT_REGEX.test(formattedNewValue) ||
+        MATCH_ZEROS_AFTER_DECIMAL_REGEX.test(formattedNewValue)
+      ) {
+        const decimalNumberParts = formattedNewValue.split(DECIMAL_NUMBER_SEPARATOR);
         const decimalPart = decimalNumberParts[1];
-        const integerPart = decimalNumberParts[0].replace(THOUSANDTHS_SEPARATOR, "");
+        const integerPart = decimalNumberParts[0];
+
+        formattedNewValue = `${integerPart}${DECIMAL_NUMBER_SEPARATOR}${decimalPart}`;
+      }
 
-        if (decimalPart && decimalPart.length > maxFractionDigits) {
-          const trimmedDecimalPart = decimalPart.slice(0, maxFractionDigits);
+      if (maximumFractionDigits > 0) {
+        const decimalPart = newValue.split(DECIMAL_NUMBER_SEPARATOR)[1];
 
-          formattedNewValue = `${integerPart}${DECIMAL_NUMBER_SEPARATOR}${trimmedDecimalPart}`;
+        if (decimalPart && decimalPart.length === maximumFractionDigits + 1) {
+          return;
         }
       } else {
```

Comment:
I think we can provide `maximumFractionDigits` to `parseNumber` as an option and handle this there.